---
title: "ExposureVisual"
author: 
  name: "Veronica Erconi"
  affiliation: Politecnico di Milano, Italy
  email: erconi.veronica@gmail.com
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{ExposureVisual}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r style, echo = FALSE, results = 'asis'}
library(BiocStyle)
```

```{r, echo = FALSE}
library(knitr)
```

```{r setup}
library(ExposureVisual)
```

#Introduction

**ExposureVisual** R package has been developed to visualize as a heatmap 
the contribution of each mutational signature to the load of somatic single 
nucleotide variants (SNVs) present in a tumor genome provided in MPF format.

To obtain the heatmap this function first decompose individual
tumor genomes (i.e., the lists of somatic SNVs identified in individual tumors) according to a set of given mutational signatures using `readGenomesFromMPF()`
function from **decompTumor2Sig** package by Kr√ºger S, Piro RM (BMC Bioinformatics 20(Suppl 4):152, 2019. See `r Biocpkg("decompTumor2Sig")`).
Then, for each of the given mutational signatures/processes, this function uses **decompTumor2Sig** `decomposeTumorGenomes()` function to determine the signatures
contributions, also called exposures, to the load of somatic mutations observed 
in a tumor genome. These exposures are finally scaled, hierarchically clustered
using correlation and complete linkage, scaled and finally visualized as heatmap
using `pheatmap()` function from **pheatmap** package (see `r CRANpkg("pheatmap")`).

**Please read the following important notes first**:

*Note*: given the number of parameters to be estimated, **ExposureVisual**
should best be applied only to tumor samples with a sufficient number of somatic
mutations. From our experience, using tumor samples with 100+ somatic mutations
was reasonable when using version 2 of the COSMIC Mutational Signatures
(30 signatures). Tumor samples with fewer somatic mutations should be avoided
unless a tumor-type specific subset of signatures is used.

*Note*: be aware that publicly available signature sets have often been defined
with respect to genome wide mutation frequencies, so they should best be applied
to somatic mutation data from whole genome sequencing. Background frequencies
may be different for subsets of the genome, i.e., current signature sets might
yield incorrect results when applied to, for example, mutation data from
targetted sequencing of only a subset of genes!

# Input data {Input}

**ExposureVisual** works with a set of SNVs observed in a tumor genome in *MPF*
format and a set of given *mutational signatures*.
In addition, the reference genome sequence (in form of a
`r Rpackage("BSgenome")` object) is required to determine neighboring nucleotides 
of the mutated bases. It may also require transcript annotations (in form of a
`r Rpackage("TxDb")` object) in case the given mutational signatures take
information on the transcription direction into account.

## Tumor genomes SNVs in MPF format {#MPF-Input}

Mutation Position Format (MPF) files are tab-delimited text files characterized 
by one line per mutation and patient or sample as follows:

[sampleID]<tab>[chrom]<tab>[position]<tab>[ref_bases]<tab>[alt_bases]

```
patient1   chr1   809687    G   C
patient1   chr1   819245    G   T
patient1   chr2   2818266   A   G
patient1   chr2   3433314   G   A
patient2   chr1   2927666   A   G
patient2   chr1   3359791   C   T
```
With:
* **sampleID** = the name of the sample (or tumor ID); 
* **chrom** = the chromosome name; 
* **position** = the position on the chromosome; 
* **ref_bases** = the reference base at that position (A, C, G, or T); 
* **alt_bases** = the alternate or variant base (A, C, G, or T).


## Mutational signatures {#Sign-input}

*Mutational signatures* represents mutational processes like UV light, smoking, etc. (see  Alexandrov and Stratton, Curr Opin Genet Dev 24:52-60, 2014) that can be 
either of the form initially proposed by Alexandrov et al. (Cell Rep. 3:246--259, 
2013 and Nature 500:415--421, 2013)---in the following called "Alexandrov signatures"---or of the simplified form proposed by Shiraishi et al. (PLoS Genet. 11:e1005657, 2015)---in the following called "Shiraishi signatures".
To easily retrieve them, we suggest using **decompTumor2Sig** specific functions
as explained at the following link: `r Biocpkg("decompTumor2Sig")`.

# ExposureHeatVisual

`ExposureHeatVisual()` is the only function of **ExposureVisual** package.
Given an input MPF file containing the somatic SNVs identified in individual tumor genomes (see \@ref(MPF) section) and a set of mutational signatures (see \@ref(Sign-input) section) it allows to determine the exposures, indicating the fractions or percentages of somatic mutations which can likely be
attributed to each of the given mutational signature in each tumor, to cluster the resulting exposure vectors and to visualize them, i.e. the strength of the contribution of the different signatures to the overall mutational load observed in each tumor, as a heatmap.

## ExposureHeatVisual usage

```{r, eval=FALSE}
ExposureHeatVisual (MPFfile,  type="Alexandrov", numBases=3,
                refGenome=BSgenome.Hsapiens.UCSC.hg19::BSgenome.Hsapiens.UCSC.hg19,
                transcriptAnno=
                TxDb.Hsapiens.UCSC.hg19.knownGene::TxDb.Hsapiens.UCSC.hg19.knownGene,
                trDir=FALSE, enforceUniqueTrDir=TRUE, signatures,
                heat_main = "Exposure Heatmap",
                cluster_rows=TRUE, cluster_cols=FALSE, cellwidth= 15, cellheight=15)

```

The steps of:
* Read of somatic mutations of tumor genomes from MPF file;
* Decomposition of tumor genomes by signature refitting;
* Exposure vectors clusterization and visualization as a heatmap;
can be controlled by the following arguments:

* *type*: Model of signatures that will be used with the genomes, "Shiraishi" or
"Alexandrov" (default).
* *numBases*: The odd Total number of bases (mutated base and flanking bases) to be used for sequence patterns. Default: 3. 
* *trDir*: Whether the transcription direction should be taken into account
(default: TRUE). If so, only mutations located within genomic regions for which
a transcript direction is defined will be considered.
* *refGenome*: The reference genom needed to extract sequence patterns. 
* *transcriptAnno*: The transcript annotation database from which to obtain the
transcription direction (if needed, i.e., if *trDir*=TRUE).
* *enforceUniqueTrDir*: If *trDir* is TRUE, then by default each mutation which
maps to a region with multiple overlapping genes with opposing transcription
directions will be excluded from the analysis. This is because from mutation
data alone it cannot be inferred which of the two genes has the higher
transcription activity which might potentially be linked to the occurrence of
the mutation. Until version 1.3.5 of **decompTumor2Sig** the behavior for
mutations associated with two valid transcription directions was different: the
transcript direction encountered first in the transcript database (specified
with *transcriptAnno*) was assigned to the mutation; the latter is also the
default behavior of the `r Rpackage("pmsignature")` package. If you need to
reproduce the old behavior---which basically arbitrarily assigns one of the two
transcriptions strands---then you can set *enforceUniqueTrDir*=FALSE (this
option exists mostly for backward compatibility with older versions), but it is
recommended to entirely exclude mutations without ambiguous transcription
strands. Note: this option is ignored when *trDir*=FALSE, where all mutations
can be used.
* *heat_main*: A string representing the title of the heatmap,"Exposure Heatmap" (default).
* *cluster_rows*: boolean value determining if rows should be clustered or hclust object (default: TRUE).
* *cluster_cols*: similar to cluster_rows but for columns (default: FALSE).
* *cellwidth*: individual cell width in points. If NA, then the values depend on the size of plotting window, if not specified, the width will be of 15 (default).
* *cellheight*: similar to cellwidth but for cell height.

## ExposureHeatVisual output

As result, the function will return the heatmap of the exposure vectors.
On the lower part we will have signature numbers in ascending order or clustered, based on the visualization needs. On the right we will have tumor/sample IDs and
the color bar. On the upper part, the chosen title will be showed.

# Example Workflow

## Alexandrov type signatures

As an example, we use the *MPF file* provided by the **decompTumor2Sig** package:
* somatic mutations for six of the 21 breast cancer genomes originally published 
by Nik-Zainal et al (Cell 149:979--993, 2012);
* version 2 Alexandrov *signatures* directly from COSMIC.
 
```{r}

sigfile <- system.file("extdata", "signatures_probabilities.txt",
                       package="ExposureVisual")

### Get Alexandrov signatures
signatures <- decompTumor2Sig::readAlexandrovSignatures(sigfile)

### Load Reference Genome
refGenome <- BSgenome.Hsapiens.UCSC.hg19::BSgenome.Hsapiens.UCSC.hg19

### Obtain MPF file (.gz in this case)
 MPFfile <- system.file("extdata", "NikZainal_mutdata_21breast.gz",
                      package="ExposureVisual")

### Compute exposures and visualize as a heatmap
ExposureHeatVisual(MPFfile, refGenome = refGenome, signatures, 
                   heat_main="Exposure Heatmap Alexandrov")

```


## Shirashi type signatures

In case of Shirashi, this example uses a small set of four Shiraishi-type 
signatures in flat files provided by **decompTumor2Sig** package.

```{r}

### Get example signature flat files provided by decompTumor2Sig
sigfiles <- system.file("extdata",
                 paste0("Nik-Zainal_PMID_22608084-pmsignature-sig",1:4,".tsv"),
                 package="decompTumor2Sig")

### Read signatures flat files
signatures <- readShiraishiSignatures(files=sigfiles)

### Load Reference Genome and the transcript annotation database
refGenome <- BSgenome.Hsapiens.UCSC.hg19::BSgenome.Hsapiens.UCSC.hg19
transcriptAnno <-
           TxDb.Hsapiens.UCSC.hg19.knownGene::TxDb.Hsapiens.UCSC.hg19.knownGene

### Obtain MPF file (.gz in this case)
 MPFfile <- system.file("extdata", "NikZainal_mutdata_21breast.txt.gz",
                      package="ExposureVisual")

### Compute exposures and visualize as a heatmap
ExposureHeatVisual(MPFfile, numBases=5, type="Shirashi", trDir= TRUE, 
                   refGenome = refGenome, transcriptAnno=transcriptAnno,
                   signatures, heat_main="Exposure Heatmap Shirashi")

```

For other approaches concerning signatures, please see `r Biocpkg("decompTumor2Sig")` vignettes.
